### ISSUE TITLE
If rule has erroneous options should it be skipped?

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#338]
Right now if I pass an erroneous option I get an output (from postcss-reporter) like this:

```
test/forVisual.css
⚠  Invalid option value "tlab" for rule "indentation" [stylelint]
2:3	⚠  Expected single space after ":" (declaration-colon-space-after) [stylelint]
2:3	⚠  Unexpected unit on zero length number (number-zero-length-no-unit) [stylelint]
2:3	⚠  Expected indentation of NaN spaces at line 2 (indentation) [stylelint]
5:3	⚠  Expected indentation of NaN spaces at line 5 (indentation) [stylelint]
8:3	⚠  Expected single space after ":" (declaration-colon-space-after) [stylelint]
8:3	⚠  Expected indentation of NaN spaces at line 8 (indentation) [stylelint]
```

All of the `Expected indentation of NaN` are because of the typo `"tlab"` instead of `"tab"`.

That's awkward.

So ... what do you think? If there's an option mistake return early from the rule?

**Comment by jeddy3** [comment]
> That's awkward.

Yes it is :)

> If there's an option mistake return early from the rule?

Yes, I think so. If only because an erroneous option, especially for something like `indentation`, could produce a ton of invalid warnings that could drown-out the original "Invalid option..." warning.

**Comment by davidtheclark** [comment]
Here's what I'm thinking ...

First of all, we change `validateOptions` API so that it can accept multiple actual/possible pairs. We could change this:

```js
validateOptions({ result, ruleName,
  actual: expectation,
  possible: [ "always", "never" ],
})
validateOptions({ result, ruleName,
  actual: options,
  possible: {
    except: [ "blockless-group", "first-nested", "all-nested" ],
    ignore: ["all-nested"],
  },
})
```

To this:

```js
validateOptions(result, ruleName, 
  {
    actual: expectation,
    possible: [ "always", "never" ],
  },
  {
    actual: options,
    possible: {
      except: [ "blockless-group", "first-nested", "all-nested" ],
      ignore: ["all-nested"],
    }
)
```

The reason for that would be so we have only one `validateOptions()` call per rule.

Then, we make `validateOptions()` return a boolean: `true` if they were valid, `false` if not. And the rules return early if it's false. E.g.

```js
return (root, result) => {
  var validOptions = validateOptions( .. )
  if (!validOptions) { return }
}
```

It would mean some annoying refactoring of many places, but it would work.

Thoughts?

**Comment by jeddy3** [comment]
It certainly feels clean and explicit. 

What can I do to help?

**Comment by davidtheclark** [comment]
Fixed by #339.
